beacon_node/beacon_chain/src/block_times_cache.rs

```diff
@@ -0,0 +1,143 @@
+        BlockDelays {
+            observed,
+            imported,
+            set_as_head,
+        }
+    }
+}
+
+// If the block was received via gossip, we can record the client type of the peer which sent us
+// the block.
+#[derive(Clone, Default)]
+pub struct BlockPeerInfo {
+    pub id: Option<String>,
+    pub client: Option<String>,
+}
+
+pub struct BlockTimesCacheValue {
+    pub slot: Slot,
+    pub timestamps: Timestamps,
+    pub peer_info: BlockPeerInfo,
+}
+
+impl BlockTimesCacheValue {
+    fn new(slot: Slot) -> Self {
+        BlockTimesCacheValue {
+            slot,
+            timestamps: Default::default(),
+            peer_info: Default::default(),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct BlockTimesCache {
+    pub cache: HashMap<BlockRoot, BlockTimesCacheValue>,
+}
+
+/// Helper methods to read from and write to the cache.
+impl BlockTimesCache {
+    pub fn set_time_observed(
+        &mut self,
+        block_root: BlockRoot,
+        slot: Slot,
+        timestamp: Duration,
+        peer_id: Option<String>,
+        peer_client: Option<String>,
+    ) {
+        let block_times = self
+            .cache
+            .entry(block_root)
+            .or_insert_with(|| BlockTimesCacheValue::new(slot));
+        block_times.timestamps.observed = Some(timestamp);
+        block_times.peer_info = BlockPeerInfo {
+            id: peer_id,
+            client: peer_client,
+        };
+    }
+
+    pub fn set_time_imported(&mut self, block_root: BlockRoot, slot: Slot, timestamp: Duration) {
+        let block_times = self
+            .cache
+            .entry(block_root)
+            .or_insert_with(|| BlockTimesCacheValue::new(slot));
+        block_times.timestamps.imported = Some(timestamp);
+    }
+
+    pub fn set_time_set_as_head(&mut self, block_root: BlockRoot, slot: Slot, timestamp: Duration) {
+        let block_times = self
+            .cache
+            .entry(block_root)
+            .or_insert_with(|| BlockTimesCacheValue::new(slot));
+        block_times.timestamps.set_as_head = Some(timestamp);
+    }
+
+    pub fn get_block_delays(
+        &self,
+        block_root: BlockRoot,
+        slot_start_time: Duration,
+    ) -> BlockDelays {
+        if let Some(block_times) = self.cache.get(&block_root) {
+            BlockDelays::new(block_times.timestamps.clone(), slot_start_time)
+        } else {
+            BlockDelays::default()
+        }
+    }
+
+    pub fn get_peer_info(&self, block_root: BlockRoot) -> BlockPeerInfo {
+        if let Some(block_info) = self.cache.get(&block_root) {
+            block_info.peer_info.clone()
+        } else {
+            BlockPeerInfo::default()
+        }
+    }
+
+    // Prune the cache to only store the most recent 2 epochs.
+    pub fn prune(&mut self, current_slot: Slot) {
+        self.cache
+            .retain(|_, cache| cache.slot < current_slot.saturating_sub(64_u64));
```

Comment:
I _think_ this is backwards... we want to keep (`retain`) new entries with slots greater than `current_slot - 64`

```suggestion
            .retain(|_, cache| cache.slot > current_slot.saturating_sub(64_u64));
```

Reply:
Yep, you're right, nice catch again :+1: